lib/tcpip/tcpip_error_handler.h

```diff
@@ -0,0 +1,367 @@
+	//    restarting
+	// 3. reset the network thread whenever it wakes up
+	uint8_t expected = NotRestarting;
+	if (!restartState.compare_exchange_strong(expected, Restarting))
+	{
+		// `expected` now contains a snapshot of `restartState`.
+		if (isIpThread && ((expected & IpThreadKicked) != 0))
+		{
+			// Currently recovering from a crash that happens
+			// during the reset process is not possible. It is not
+			// clear if we ever really want to do that: we will
+			// only crash during reset if 1) there is a bug in the
+			// reset code, or 2) there is some global data that we
+			// cannot reset and which is corrupted. In either case,
+			// re-reseting the same way will not make the situation
+			// better.
+			DebugErrorHandler::log("The network thread crashed while "
+			                       "restarting. This may be unrecoverable.");
+		}
+
+		// Another instance of the error handler is running, do not do
+		// anything.
+		return;
+	}
+
+	DebugErrorHandler::log("Reset-ing the network stack.");
+
+	// We need to acquire the sealed sockets lock because we do not want
+	// the sealed sockets list to be in an inconsistent state when we go
+	// over it.
+	//
+	// Waiting to acquire the lock is fine, as we know that any thread
+	// which holds the it will eventually release it, either 1) exiting the
+	// critical section, or 2) crashing into it, in which case we (the
+	// error handler) will manually unlock it (see manual unlock above).
+	DebugErrorHandler::log("Acquiring the sealed sockets lock.");
+	sealedSocketsListLock.lock();
+
+	DebugErrorHandler::log(
+	  "Setting the sealed sockets list lock for destruction.");
+	sealedSocketsListLock.upgrade_for_destruction();
+
+	// Upgrade socket locks for destruction and destroy event groups to
+	// ensure that threads waiting on them exit the compartment. We will
+	// empty the list later.
+	// TODO here we may want to check if the ring is well-formed
+	DebugErrorHandler::log(
+	  "Setting socket locks for destruction and destroying event groups.");
+	if (!sealedSockets.is_empty())
+	{
+		auto *cell = sealedSockets.first();
+		while (cell != &sealedSockets.sentinel)
+		{
+			struct SealedSocket *socket = SealedSocket::from_ring(cell);
+
+			FlagLockPriorityInherited *lock = &(socket->socketLock);
+			if (Capability{lock}.is_valid())
+			{
+				DebugErrorHandler::log("Destroying socket lock {}.", lock);
+				lock->upgrade_for_destruction();
+			}
+			else
+			{
+				DebugErrorHandler::log("Ignoring corrupted socket lock {}.",
+				                       lock);
+			}
+
+			FreeRTOS_Socket_t *s = socket->socket;
+			if (Capability{s}.is_valid() &&
+			    Capability{s->xEventGroup}.is_valid())
+			{
+				DebugErrorHandler::log("Destroying event group {}.",
+				                       s->xEventGroup);
+				eventgroup_destroy_force(MALLOC_CAPABILITY, s->xEventGroup);
+			}
+			else
+			{
+				DebugErrorHandler::log("Ignoring corrupted socket {}.", s);
```

Comment:
Presumably memory for the event group will be cleaned up by the `heap_free_all` later, even if the pointer to it is corrupted?

Reply:
Right, the memory of the event group will still be freed later with the `heap_free_all`, however we run the risk to have the IP thread stuck on the event queue which we didn't manage to destroy.

Reply:
I added a comment to clarify this.